cardano-testnet/test/Test/Cli/Babbage/LeadershipSchedule.hs

```diff
@@ -0,0 +1,194 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+{- HLINT ignore "Redundant id" -}
+{- HLINT ignore "Redundant return" -}
+{- HLINT ignore "Use head" -}
+{- HLINT ignore "Use let" -}
+
+module Test.Cli.Babbage.LeadershipSchedule
+  ( hprop_leadershipSchedule
+  ) where
+
+import           Cardano.CLI.Shelley.Output (QueryTipLocalStateOutput (..))
+import           Control.Monad (void)
+import           Data.List ((\\))
+import           Data.Monoid (Last (..))
+import           GHC.Stack (callStack)
+import           Hedgehog (Property)
+import           Prelude
+import           System.Environment (getEnvironment)
+import           System.FilePath ((</>))
+
+import qualified Data.Aeson as J
+import qualified Data.Aeson.Types as J
+import qualified Data.List as L
+import qualified Data.Time.Clock as DTC
+import qualified Hedgehog as H
+import qualified Hedgehog.Extras.Stock.IO.Network.Sprocket as IO
+import qualified Hedgehog.Extras.Test.Base as H
+import qualified Hedgehog.Extras.Test.File as H
+import qualified Hedgehog.Extras.Test.Process as H
+import qualified System.Directory as IO
+import qualified System.Info as SYS
+import qualified Testnet.Util.Base as H
+
+import           Cardano.Testnet
+import           Testnet.Util.Assert
+import           Testnet.Util.Process
+import           Testnet.Util.Runtime
+
+hprop_leadershipSchedule :: Property
+hprop_leadershipSchedule = H.integrationRetryWorkspace 2 "babbage-leadership-schedule" $ \tempAbsBasePath' -> do
+  H.note_ SYS.os
+  base <- H.note =<< H.noteIO . IO.canonicalizePath =<< H.getProjectBase
+  configurationTemplate <- H.noteShow $ base </> "configuration/defaults/byron-mainnet/configuration.yaml"
+  conf@Conf { tempBaseAbsPath, tempAbsPath } <- H.noteShowM $
+    mkConf (ProjectBase base) (YamlFilePath configurationTemplate) tempAbsBasePath' Nothing
+
+  work <- H.note $ tempAbsPath </> "work"
+  H.createDirectoryIfMissing work
```

Comment:
There are some convenience functions provided by `hedgehog-extras`.  These functions correspond to the ones found in standard Haskell libraries.

`createDirectoryIfMissing` for example is the `hedgehog-extras` version of the one in `System.Directory`.  The difference is that this one is "exception-safe" and also annotates the output.

`createDirectoryIfMissing` is implemented like this:

```
createDirectoryIfMissing :: (MonadTest m, MonadIO m, HasCallStack) => FilePath -> m ()
createDirectoryIfMissing filePath = GHC.withFrozenCallStack $ do
  H.annotate $ "Creating directory if missing: " <> filePath
  H.evalIO $ IO.createDirectoryIfMissing True filePath
```

`annotate` is the same as `note`.  `evalIO` calls the IO action in an "exception-safe" manner.  It is the same as `noteIO` except it doesn't annotate.  `noteIO` is implemented in terms of `evalIO`.

The call to `withFrozenCallStack` ensures that annotations are attached to the caller, not here.  For conveniences functions like these, it is advised to always do this because we care about the annotations in the context of the failing test, not the convenience function.  To make this work we also need to use the `HasCallStack` constraint.

This is the output of that line:

```
             56 ┃   H.createDirectoryIfMissing work
                ┃   │ Creating directory if missing: /private/tmp/nix-shell.0QhPpd/babbage-leadership-schedule-0-test-55125045e4f9b46e/work
```